pkg/koordlet/metricsadvisor/collector_gpu.go

```diff
@@ -0,0 +1,269 @@
+		}
+
+		uuid, ret := gpudevice.GetUUID()
+		if ret != nvml.SUCCESS {
+			return fmt.Errorf("unable to get device uuid: %v", nvml.ErrorString(ret))
+		}
+
+		minor, ret := gpudevice.GetMinorNumber()
+		if ret != nvml.SUCCESS {
+			return fmt.Errorf("unable to get device minor number: %v", nvml.ErrorString(ret))
+		}
+		devices[deviceIndex] = &device{
+			DeviceUUID: uuid,
+			Minor:      int32(minor),
+			Device:     gpudevice,
+		}
+	}
+
+	g.Lock()
+	defer g.Unlock()
+	g.deviceCount = count
+	g.devices = devices
+	return nil
+}
+
+func (g *gpuDeviceManager) getNodeGPUUsage() []metriccache.GPUMetric {
+	g.RLock()
+	defer g.RUnlock()
+	tmp := make([]rawGPUMetric, g.deviceCount)
+	for i := 0; i < g.deviceCount; i++ {
+		tmp[i] = rawGPUMetric{}
+	}
+	for _, p := range g.processesMetrics {
+		for idx := 0; idx < g.deviceCount; idx++ {
+			if m := p[uint32(idx)]; m != nil {
+				tmp[idx].SMUtil += p[uint32(idx)].SMUtil
+				tmp[idx].MemoryUsed += p[uint32(idx)].MemoryUsed
+			}
+		}
+	}
+	rtn := make([]metriccache.GPUMetric, g.deviceCount)
+	for i := 0; i < g.deviceCount; i++ {
+		rtn[i] = metriccache.GPUMetric{
+			DeviceUUID: g.devices[i].DeviceUUID,
+			Minor:      g.devices[i].Minor,
+			SMUtil:     tmp[i].SMUtil,
+			MemoryUsed: *resource.NewQuantity(int64(tmp[i].MemoryUsed), resource.BinarySI),
+		}
+	}
+	return rtn
+}
+
+func (g *gpuDeviceManager) getPIDsTotalGPUUsage(pids []uint64) []metriccache.GPUMetric {
+	g.RLock()
+	defer g.RUnlock()
+	tmp := make(map[int]*rawGPUMetric)
+	for _, pid := range pids {
+		if metrics, exist := g.processesMetrics[uint32(pid)]; exist {
+			for idx, metric := range metrics {
+				if metric == nil {
+					continue
+				}
+				if _, found := tmp[idx]; !found {
+					tmp[idx] = &rawGPUMetric{}
+				}
+				tmp[idx].MemoryUsed += metric.MemoryUsed
+				tmp[idx].SMUtil += metric.SMUtil
+			}
+		}
+	}
+	if len(tmp) == 0 {
+		return nil
+	}
+	rtn := make([]metriccache.GPUMetric, 0)
+	for i := 0; i < g.deviceCount; i++ {
+		if value, ok := tmp[i]; ok {
+			rtn = append(rtn, metriccache.GPUMetric{
+				DeviceUUID: g.devices[i].DeviceUUID,
+				Minor:      g.devices[i].Minor,
+				SMUtil:     value.SMUtil,
+				MemoryUsed: *resource.NewQuantity(int64(value.MemoryUsed), resource.BinarySI),
+			})
+		}
+	}
+	return rtn
+}
+
+func (g *gpuDeviceManager) getPodGPUUsage(podParentDir string, cs []corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	pids, err := util.GetPIDsInPod(podParentDir, cs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pid, error: %v", err)
+	}
+	return g.getPIDsTotalGPUUsage(pids), nil
+}
+
+func (g *gpuDeviceManager) getContainerGPUUsage(podParentDir string, c *corev1.ContainerStatus) ([]metriccache.GPUMetric, error) {
+	currentPIDs, err := util.GetPIDsInContainer(podParentDir, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pid, error: %v", err)
+	}
+	return g.getPIDsTotalGPUUsage(currentPIDs), nil
+}
+
+func (g *gpuDeviceManager) collectGPUUsage() {
+	g.RLock()
```

Comment:
g.RLock() for g.deveices is not necessary, since the write only happens during init.